### PR TITLE
Fix `parametrize(... ids=<function>)` when the function returns non-strings

### LINKED ISSUE #4739: `ids=func` where `func` doesn't return strings results in internal error
```python
import pytest


def id_gen(dct):
    return dct['k']


@pytest.mark.parametrize(
    'f',
    ({'k': {'val': 'val'}}, {'k': {'val2', 'val2'}}),
    ids=id_gen,
)
def test(f):
    assert f
```

```
$ pytest t.py 
============================= test session starts ==============================
platform linux -- Python 3.6.7, pytest-4.2.0, py-1.7.0, pluggy-0.8.1
rootdir: /tmp/x, inifile:
collected 0 items / 1 errors                                                   

==================================== ERRORS ====================================
____________________________ ERROR collecting t.py _____________________________
venv/lib/python3.6/site-packages/pluggy/hooks.py:284: in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
venv/lib/python3.6/site-packages/pluggy/manager.py:68: in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
venv/lib/python3.6/site-packages/pluggy/manager.py:62: in <lambda>
    firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
venv/lib/python3.6/site-packages/_pytest/python.py:226: in pytest_pycollect_makeitem
    res = list(collector._genfunctions(name, obj))
venv/lib/python3.6/site-packages/_pytest/python.py:403: in _genfunctions
    self.ihook.pytest_generate_tests(metafunc=metafunc)
venv/lib/python3.6/site-packages/pluggy/hooks.py:284: in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
venv/lib/python3.6/site-packages/pluggy/manager.py:68: in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
venv/lib/python3.6/site-packages/pluggy/manager.py:62: in <lambda>
    firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
venv/lib/python3.6/site-packages/_pytest/python.py:133: in pytest_generate_tests
    metafunc.parametrize(*marker.args, **marker.kwargs)
venv/lib/python3.6/site-packages/_pytest/python.py:996: in parametrize
    ids = self._resolve_arg_ids(argnames, ids, parameters, item=self.definition)
venv/lib/python3.6/site-packages/_pytest/python.py:1050: in _resolve_arg_ids
    ids = idmaker(argnames, parameters, idfn, ids, self.config, item=item)
venv/lib/python3.6/site-packages/_pytest/python.py:1196: in idmaker
    for valindex, parameterset in enumerate(parametersets)
venv/lib/python3.6/site-packages/_pytest/python.py:1196: in <listcomp>
    for valindex, parameterset in enumerate(parametersets)
venv/lib/python3.6/site-packages/_pytest/python.py:1186: in _idvalset
    for val, argname in zip(parameterset.values, argnames)
venv/lib/python3.6/site-packages/_pytest/python.py:1186: in <listcomp>
    for val, argname in zip(parameterset.values, argnames)
venv/lib/python3.6/site-packages/_pytest/python.py:1158: in _idval
    return ascii_escaped(s)
venv/lib/python3.6/site-packages/_pytest/compat.py:239: in ascii_escaped
    ret = val.encode("unicode_escape").decode("ascii")
E   AttributeError: 'dict' object has no attribute 'encode'
!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!!
=========================== 1 error in 0.17 seconds ============================
```

This apparently used to work in 3.0.7:

```
$ pytest t.py  -vvv
============================= test session starts ==============================
platform linux -- Python 3.6.7, pytest-3.0.7, py-1.7.0, pluggy-0.4.0 -- /tmp/x/venv/bin/python3
cachedir: .cache
rootdir: /tmp/x, inifile:
collected 2 items 

t.py::test[f0] PASSED
t.py::test[f1] PASSED

=========================== 2 passed in 0.00 seconds ===========================
```

### PR DESCRIPTION
**Pull request by asottile**
Resolves #4739

### CODE REVIEW

File: src/_pytest/python.py [hunk 1]
```diff
@@ -1144,19 +1144,17 @@ def _find_parametrized_scope(argnames, arg2fixturedefs, indirect):
...
+                val = generated_id
         except Exception as e:
             # See issue https://github.com/pytest-dev/pytest/issues/2169
             msg = "{}: error raised while trying to determine id of parameter '{}' at position {}\n"
             msg = msg.format(item.nodeid, argname, idx)
             # we only append the exception type and message because on Python 2 reraise does nothing
             msg += "  {}: {}\n".format(type(e).__name__, e)
             six.raise_from(ValueError(msg), e)
-        if s:
-            return ascii_escaped(s)
```
Comment (RonnyPfannschmidt) [inline src/_pytest/python.py]: this is a behavior change, the hook now can override the given id
additionally i noted we font factor in the `ascii_escape` of the id from the hook

i propose factoring out the escaping logic to a function that gets a fallback result and using that to wrap around the return value in the different cases

Comment (asottile) [reply src/_pytest/python.py]: _headscratch_ I'm not seeing the behaviour change -- can you shed some light?  The only behaviour I intended to change is crash -> not crash (and restore what happened in pytest 3.0.7)

Comment (RonnyPfannschmidt) [reply src/_pytest/python.py]: by moving the idfn handling from returning before the `pytest_make_parametrize_id` to after invoking it,
the behavior is accidentally changed, as now the hook can override the result of the `idfn`, while before it couldn't

Comment (asottile) [reply src/_pytest/python.py]: That sounds like an improvement, no?

Comment (RonnyPfannschmidt) [reply src/_pytest/python.py]: as done here its possible that this inadvertely changes ids in a testsuite, so its a change for a major release

Comment (asottile) [reply src/_pytest/python.py]: it's easy to just `elif` I guess

Comment (asottile) [reply src/_pytest/python.py]: seems broken though that in this weird special case we don't apply the hook

Comment (RonnyPfannschmidt) [reply src/_pytest/python.py]: It is broken, its just that bugfixing is a breaking behavioral change

Comment (asottile) [reply src/_pytest/python.py]: how if fixing any bug not a behavioural change :rofl:

Comment (asottile) [reply src/_pytest/python.py]: @RonnyPfannschmidt better now?

Comment (RonnyPfannschmidt) [reply src/_pytest/python.py]: it seems much better, my gut is not 100% happy with the solution we have tho,
i'd really like to note down to change the behavior for pytest 6.0 (avoiding it in 5.0 for the management of the pytohn support drop)

**Comment by RonnyPfannschmidt** [comment]
@asottile would you like to create the followup issue, otherwise i will

**Comment by asottile** [comment]
> 
> 
> @asottile would you like to create the followup issue, otherwise i will

I think you'll do it better justice than I will 👍

**Comment by RonnyPfannschmidt** [comment]
thanks, i'm on it then

**Pull request merged**